### ISSUE TITLE
Extending profiles for filters, sorting, paging, ranges, etc.

### ISSUE THREAD

**Issue by garysweaver** [json-api/json-api#45]
To start off, I need to describe what we have already; we have a set of URL conventions used in: https://github.com/rubyservices/restful_json including (though there are others):

filter:

```
https://example.org/foobars?foo_id=123
```

filter greater than:

```
https://example.org/foobars?bar_date!gt=2012-08-08
```

filter greater than and return record count:

```
https://example.org/foobars?bar_date!gt=2012-08-08&count=
```

filter on "named attribute" (not a real resource field- represents a field of a direct or indirect relation defined server-side) and return unique (distinct) records:

```
https://example.org/foobars?and_one_more=an_attribute_value_on_my_assocs_assocs_assoc&uniq=
```

using the server-defined page size, return page count:

```
https://example.org/foobars?page_count=
```

return page at specific number, using server-determined page size:

```
https://example.org/foobars?page=1
```

range/more flexible paging:

```
https://example.org/foobars?skip=30&take=15
```

There are good reasons _not_ to try to promote generalized querying for a variety of security and performance reasons, which is why we try to only provide each ability of the controller when the developer enabled that feature.

These are common relational DB concepts, so when each _is_ enabled, it would be nice to have a standard agreed upon way of constructing a URL like used in that project. The ARel syntax used may not be optimal and could conflict with attribute/column names and require work in the model layer in those rare cases, but it works as a first cut.

The following http://jsonapi.org/extending/ provides an example for paging and it sounds like more examples of metas/profiles for the spec are under construction.

So, some questions:
1. Is there going to be an official list of methods for defining a common subset of all of the extensions (profiles) as suggestions? It would be great to get people to agree on filtering, etc. similar to what we have above or to change our framework to work like some agreed-upon spec. They aren't going to keep adding them all to section 5 of [RFC 6906](http://tools.ietf.org/html/rfc6906). i.e. should there be a place to list common profiles/meta URLs, and how would that be administered?
2. How do we avoid bloat of N metas/profiles in each response? Should things like all of the above features (filtering, paging, etc.) be a single profile URL (a group) or individual meta URLs?
3. Is there going to be a suggested best practice for profile URLs to contain a version number in the profile URL? Could it be suggested to allow rational versions and suggest that client and server frameworks support partial matching of those versions, similar to gem versioning?: http://docs.rubygems.org/read/chapter/7 That way, enhanced versions of profiles could be provided where the URL of the profile changes, but the client could know that what they are doing is still compatibile, e.g.:

The client originally was developed to work with v1.2.3 of the profile `http://api.example.com/simple-querying` which came back at the time with:

```
{
  "meta": {
    "profile": "http://api.example.com/simple-querying?v=1.2.3"
  },
  ...
}
```

But later the API is changed slightly to fix a bug in a way that is still compatible with v1.2 of the API, so it returns the profile `http://api.example.com/simple-querying?v=1.2.4` (perhaps some gem being used on the Rails server-side was updated). The client could continue to work without changes since it will continue to function as before for any version that starts with 1.2. This version potentially could even be the gem version itself that implements that functionality.

In summary, I'm wondering if (1) supporting extension for querying, etc. such as I've described above will bloat the response with profile URLs in the meta section, (2) whether there will be centralized administration to suggest commonly used profiles (like those used for querying as mentioned above, or perhaps some other standard for it), and (3) whether a rational profile versioning policy for profiles should be explicitly stated in the spec.

Thanks!

**Comment by steveklabnik** [comment]
> Is there going to be an official list of methods for defining a common subset of all of the extensions (profiles) as suggestions?

Probably not. I need to re-read the profile RFC, but I think you can only have one?

> should there be a place to list common profiles/meta URLs,

Ideally, profiles include the _application semantics_, so really, there shouldn't be a lot of commonality.

> How do we avoid bloat of N metas/profiles in each response? 

Anything which is used by a large number of people should become an actual addition to the spec. Think of it as 'pull up method', the refactoring technique.

> Is there going to be a suggested best practice for profile URLs to contain a version number in the profile URL? 

Nope. In general, media types generally aren't versioned. There's nothing stopping you from including a version somehow if you'd like; for example, you could namespace your profiles to introduce one (`"profile":"http://mysite.com/v2/profile"`).

I _think_ that covers everything. Let me know if I didn't.

**Comment by garysweaver** [comment]
> > should there be a place to list common profiles/meta URLs,
> 
> Ideally, profiles include the application semantics, so really, there shouldn't be a lot of commonality.

I don't think usually that Rails apps at least are going to be describing custom profile URLs. If someone integrates with a specific api, be it their own application's api or a specific vendor's, they won't benefit as much from some URL that describes something unless it is solely for identification and version as a sanity check that the think that they integrated with did not change.

I see much more benefit in a URL that either describes a package of functionality like a gem that alters API behavior, or it describes a specific behavior or set of behaviors (like paging using a specific URL scheme).

If it describes a package of behavior it might have a 1:1 relationship between the version of the gem and the URL, i.e. `https://api.rubygems.org/restful_json/v5.0.0`, such that the restful_json gem would emit that version in its response and the client could look at it and determine whether it was compatible, and if not, it may quit or complain unless it matched/fuzzy matched it. It could even have the api description of part of gemspec and rubygems.org could host the human readable description. (And that wouldn't be Ruby specific- the same could be done for Java, etc.- they'd just need their own.) This seems cool, but I don't think it is as useful, because it isn't describing specific behaviors that the client can use, and is basically not much different than application-specific integration. It also would expose what gems a vendor is using in their API, so could even be a security hole.

If it describes a specific behavior instead, it might look like `https://some.standards.org.or.vendor/api/paging`. Multiple gems that provide paging functionality through controller api extension (like restful_json and others) could use the same URL. Then the client would know it could just append `&page=1` to the URL to get the first page of results. But, if this is done, we should really have some sort of standards organization that is ok'ing different schemes so that they were well-planned and could last longer. I think this should be done to greatly future-proof the spec. Most of the current spec could even be split into different behaviors. This would increase the size of the responses significantly, so perhaps it could be a separate single call to get a list of behaviors of the spec that could be cached in the client.

**Comment by jokeyrhyme** [comment]
Seems related to #45

**Comment by jokeyrhyme** [comment]
Oops, seems related to #21
